### ISSUE TITLE
Frontend should break up multiline selections that get sent to Console (by expression or by `\n`)

### ISSUE THREAD

**Issue by jmcphers** [posit-dev/positron#1326]
To reproduce, create a .R file that contains 50 lines, each of which has this text on it:

```R
# This is a long comment that takes a lot of space. It will be duplicated many times.
```

Then, start (or restart) R, such that the R session hasn't run any code yet. 

Highlight the entire contents of the .R file and press Cmd+Enter. The R kernel exits instead of running the code. 

<img width="1323" alt="image" src="https://github.com/posit-dev/positron/assets/470418/bd804974-f4cb-4d2e-a658-a72197d7a851">

Interesting facts:

- There is definitely a specific threshold at which this starts happening. For example, it won't reproduce if you only run 30 lines from the above test file. 
- The kernel isn't crashing! It's exiting, and even exiting normally. 

```
Executable module set to "/Users/jmcphers/git/positron/extensions/positron-r/resources/ark/ark".
Architecture set to: arm64-apple-macosx-.
(lldb) c
Process 75332 resuming
Process 75332 exited with status = 0 (0x00000000)
```

**Comment by jmcphers** [comment]
Came up again today on Bluesky: https://bsky.app/profile/tjmahr.com/post/3kwctblqx362o

```
wait i think the block was just too big. if i use a 500-line list() for my targets, i get

Error:
! Can't pass console input on to R, it exceeds R's internal console buffer size.
This is a Positron limitation we plan to fix. In the meantime, you can:
[...break it up or use source()]
```

**Comment by jennybc** [comment]
Also here a couple days ago: https://x.com/meghansharris/status/1807022600251408552

**Comment by jmcphers** [comment]
Also just reported on the discussion forum. https://github.com/posit-dev/positron/discussions/4262

**Comment by DavisVaughan** [comment]
@lionel- and I talked about this today, we have a few notes on how this should work

## Frontend

We think the frontend should be in charge of splitting large selections into _expressions_ to send to the backend.

The practical way to do this is to try and reuse some existing behavior we already have for pending input:
- Split input by `\n`
- For each line, check if that line is a complete expression with `isCodeFragmentComplete()`
- If not complete, append the next line and try again
- When you get a complete fragment, send it to the backend

https://github.com/posit-dev/positron/blob/f1d6e84242419a2594317b4338dff1658bf0b02a/src/vs/workbench/services/positronConsole/browser/positronConsoleService.ts#L1944-L1954

While this is currently being done for "pending input", it is not yet being done for the case where you send a selection of code to R while R is already idle.

In practice, this _technically_ means we could send multiple expressions to R at once, like

```r
1 + 1; 2 + 2
```

That's two expressions on 1 line. But RStudio has been doing this for years, and it really doesn't cause any issues, so I am completely fine with that behavior.

### Input / Output grouping

One important bit that comes from this change is that it will positively impact how inputs and outputs are grouped together in the Positron console. Currently it looks like this when you send a whole selection to the console:

<img width="507" alt="Screenshot 2024-08-07 at 12 00 22 PM" src="https://github.com/user-attachments/assets/1d0868d5-3bc3-46ce-839f-9ac25e42619d">

That's because Positron isn't splitting anything. It treats the whole selection as one "input" and everything we get back as one "output" that goes with that input.

But if Positron splits by expression then we will be able to replicate this much nicer RStudio behavior

<img width="275" alt="Screenshot 2024-08-07 at 12 02 03 PM" src="https://github.com/user-attachments/assets/cee382e9-c9f4-4f30-802c-0926d28b8f88">

### Expressions that are too darn long

In RStudio, the frontend actually splits by `\n`, not by expression. This means that when RStudio and R talk to each other, R gets to say "hey, this is an incomplete expression, give me more input". In Positron, since Ark is a Jupyter kernel it is typically much better if we send it a known-to-be-complete expression rather than a single line at a time.

But this means that in Positron we can run into the following case:

```r
{

 # a really big top level expression

}
```

where this 1 expression is longer than R's buffer, which brings me to....

## Backend

To support the case of really really long single expressions (which is honestly pretty common!), Ark itself is going to gain the ability to split that long complete expression by `\n`. It is way way easier and faster for Ark to do this dance with R where Ark provides one line at a time to R, R gets a chance to say "this is incomplete, give me more", and Ark's read-console hook can provide the next line. The important part is that Positron gives Ark 1 complete expression - regardless of length, Ark should be able to handle it with this change.

We think that will also improve the experience when Ark is used in other frontends besides Positron.

**Comment by jmcphers** [comment]
> If not complete, append the next line and try again
> When you get a complete fragment, send it to the backend

This does have the advantage of working w/o any additional support from language packs, but it's very chatty. In the pathological case like the `{ ... really big top level expression ... }`, we're going to be calling the backend 100 times! I think it would be worth adding a way for the frontend to ask the backend to break up the statement all at once.

The other thing worth noting is that I think the queue of statements to be run ultimately needs to be moved -- an in-memory array of pending inputs gets wiped if you close your browser, or reload the window, but I think users will expect them to continue to be consumed by the backend (especially if they are queued b/c some long-running thing is going on). Some notes on that in #1155.

**Comment by DavisVaughan** [comment]
> I think it would be worth adding a way for the frontend to ask the backend to break up the statement all at once.

Yea @lionel- thought we might need that too. Running the whole selection through R's C level `R_ParseVector()` _should_ work pretty well for this, as I think it would return a list of the individual expressions to run

**Comment by kylebutts** [comment]
Could there be a possible stop-gap measure to do something like "take current selection, write to temporary R file, and source the file"? I'm finding myself hitting this bug a lot. My current solution is to delete text that I don't want to run, sourcing the file, and then undoing the deletion (scary!)

**Comment by lionel-** [comment]
@kylebutts We're working on it!

**Comment by lionel-** [comment]
Because of the possibility that a single expression exceeds the input buffer size, we are going to prioritise fixing on the backend side: #4745.

On the frontend side, the fix will mostly be a UX improvement, in particular to get correct interpolation of inputs and outputs with multiline selections.

To support splitting inputs into complete expressions, Ark now has a new LSP request of type `positron/inputBoundaries` https://github.com/posit-dev/ark/pull/533.

Here are the types involved:

```ts
type ParseBoundaries = Array<ParseBoundary>

interface LineRange {
  start: number;
  end: number;
}

interface ParseBoundary {
  kind: 'whitespace' | 'complete' | 'incomplete' | 'invalid';
  range: LineRange;
  data: ParseBoundaryData;
}

interface ParseBoundaryData {
  [k: string]: unknown;
}

interface ParseBoundaryDataInvalid {
  message: string;
}
```

The boundaries are a vector of sorted ranges. Invariants:

- The ranges are sorted, non-overlapping, and contiguous.
- Empty lines and lines containing only whitespace or comments are complete inputs.
- There is always at least one range since the empty string is a complete input.
- Inputs are classified in complete, incomplete, and error sections. The sections are sorted in this order (there cannot be complete inputs after an incomplete or error one, there cannot be an incomplete input after an error one, and error inputs are always trailing).
- There is at most one incomplete and/or one error input in a set of inputs.

More notes:

- UI: We might not need any changes to the console UI, and instead enqueue complete expressions invisibly. Here is how it works in RStudio:

    https://github.com/user-attachments/assets/732c4ea1-8a83-4cbe-a466-9f04d0a1f000

- API: @juliasilge suggested that the best place to expose the input boundaries functionality is probably in `positron.d.ts`. Then how an extension like positron-r implements it (LSP request, Jupyter request, a process invokation) is an implementation detail.

- Fallback path: We will need to support a fallback path for runtimes that don't support an `inputBoundaries` request, in particular Python.

- When an error occurs, the input queue should be flushed, for consistency with RStudio and the backend fix being implemented in #4745.

**Comment by lionel-** [comment]
We now split by newlines on the backend side, which fixes the errors with overflowing inputs, but still need to split by expressions on the frontend side. We're now tracking the latter in #5272.
